File: week01/완전탐색/10971.py
from itertools import permutations
import sys
input = sys.stdin.readline
# 1. 입력받기
N = int(input())
permut = list(permutations(range(1, N)))   # 도시 방문 순서 순열로 완전탐색
cities = []

for i in range(N):
  city_list = list(map(int, input().split()))
  cities.append(city_list)
# 2. 최소값 할당할 변수 선언
min_cost = float('inf') # 무한대로 아싸리 큰 값 설정

# 출발지, 리스트를 인자로 받는 함수
# Cycle 도는 모든 경로 계산(순열로) => 각 경로마다 가격을 리턴
# Cycle => 출발지 => 출발지 + 모든 노드 방문

for perm in permut:
  total_cost = 0 # 각 경로별 비용
  valid = True
  
  # 출발점은 0 
  start = 0
  for next_city in perm:
    if cities[start][next_city] == 0:
      valid = False
      break
    total_cost += cities[start][next_city]
    start = next_city # 이동한 도시를 다시 출발점으로
  
  # Cycle 이니 마지막 방문도시에서 다시 돌아가야함
  if valid and cities[start][0] != 0:
    total_cost += cities[start][0]
    min_cost = min(min_cost, total_cost)

print(min_cost)

File: week01_JS/완전탐색/10971.js
const getPermutations = (arr) => {
  if (arr.length === 1) return [arr];
  const result = [];

  arr.forEach((fixed, idx, origin) => {
    const rest = [...origin.slice(0, idx), ...origin.slice(idx + 1)];
    const perms = getPermutations(rest);
    const attached = perms.map((perm) => [fixed, ...perm]);
    result.push(...attached);
  });

  return result;
};

// 1. 입력받기
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const N = parseInt(input[0]);
const cities = [];
for (let i = 1; i <= N; i++) {
  cities.push(input[i].split(' ').map(Number));
}

// 2. 모든 경로의 순열 생성
const permut = getPermutations(Array.from({ length: N - 1 }, (_, i) => i + 1));

// 3. 최소값 할당할 변수 선언
let minCost = Infinity;

// 4. 순열 순회하며 비용 계산
for (const perm of permut) {
  let totalCost = 0;
  let valid = true;
  let start = 0;

  for (const nextCity of perm) {
    if (cities[start][nextCity] === 0) {
      valid = false;
      break;
    }
    totalCost += cities[start][nextCity];
    start = nextCity;
  }

  // 순회가 유효하고, 마지막 도시에서 출발지로 돌아오는 비용이 존재하는 경우
  if (valid && cities[start][0] !== 0) {
    totalCost += cities[start][0];
    minCost = Math.min(minCost, totalCost);
  }
}

console.log(minCost);
